File: Problems/01. Easy/Contains Duplicate/Go/ContainsDuplicate.go
package main

import (
	"fmt"
	"sort"
)

// Solution 2: Sort the array and check if any adjacent elements are equal

func S2_containsDuplicate(nums []int) bool {
	// sort the slice in ascending order
	sort.Ints(nums)

	// Check for adjacent duplicates
	for i := 0; i < len(nums)-1; i++ {
		if nums[i] == nums[i+1] {
			return true
		}
	}
	return false
}

// Solution 3: Set (Hash Table)

func S3_containsDuplicate(nums []int) bool {
	hash := make(map[int]bool);

	for _, num := range nums {
		if _, exists := hash[num]; exists {
			return true;
		}
		hash[num] = true;
	}
	return false;
}

// Main

func main() {
	nums := []int{1, 5, 3, 2, 5, 4}

	if S2_containsDuplicate(nums) {
		fmt.Println("The array contains duplicates")
	} else {
		fmt.Println("The array does not contain duplicates")
	}

}

File: Problems/01. Easy/Contains Duplicate/Rust/ContainsDuplicate.rs

fn contains_dulicate(nums: &mut Vec<i32>) -> bool {
    false
}

// Solution 2: Sort the array and check if any adjacent elements are equal

fn S2_contains_duplicate(nums: &mut Vec<i32>) -> bool {
    nums.sort(); // O(n log n)

    for i in 0..nums.len() - 1 {
        if nums[i] == nums[i + 1] {
            return true
        }
    }

    false

    // you can also use slice.windows method
    // for win in nums[..].windows(2) {
    //     if win[0] == win[1] {
    //         return true
    //     }
    // }
    
}

// Solution 3: Set (Hash Table)

fn S3_contains_duplicate(nums: &[i32]) -> bool {
    use std:collectinos:HashSet;
    let mut hash = HashSet::new();

    for &num in nums.iter() {
        if hash.contains(&num) {
            return true;
        }
        hash.insert(num);
    }
    false;
}

// Main

fn main() {
    let mut nums = vec![1, 5, 3, 2, 5, 4];


    if S2_contains_duplicate(&mut nums) {
        println!("Vector contains duplicates.");
    } else {
        println!("Vector does not containe duplicates.")
    }
}


